File: box/src/main/java/de/qabel/box/storage/LocalReadBackend.java
package de.qabel.box.storage;

import de.qabel.box.storage.exceptions.QblStorageException;
import de.qabel.box.storage.exceptions.QblStorageNotFound;
import org.apache.commons.codec.digest.DigestUtils;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class LocalReadBackend implements StorageReadBackend {
    private static final Logger logger = LoggerFactory.getLogger(LocalReadBackend.class);
    private final Path root;

    public LocalReadBackend(Path root) {
        this.root = root;
    }

    @Override
    public StorageDownload download(String name) throws QblStorageException {
        try {
            return download(name, null);
        } catch (UnmodifiedException e) {
            throw new IllegalStateException(e);
        }
    }

    @Override
    public StorageDownload download(String name, String ifModifiedVersion) throws QblStorageException, UnmodifiedException {
        Path file = root.resolve(name);

        try {
            if (ifModifiedVersion != null && getMHash(file).equals(ifModifiedVersion)) {
                throw new UnmodifiedException();
            }
        } catch (IOException e) {
            // best effort
        }

        logger.info("Downloading file path " + file);
        try {
            return new StorageDownload(
                Files.newInputStream(file),
                getMHash(file),
                Files.size(file)
            );
        } catch (IOException e) {
            throw new QblStorageNotFound(e);
        }
    }

    @NotNull
    private String getMHash(Path file) throws IOException {
        return new String(DigestUtils.md5(Files.newInputStream(file)));
    }

    @Override
    public String getUrl(String meta) {
        return root.resolve(meta).toString();
    }
}

File: box/src/main/java/de/qabel/box/storage/LocalReadBackend.kt
package de.qabel.box.storage

import de.qabel.box.storage.exceptions.QblStorageException
import de.qabel.box.storage.exceptions.QblStorageNotFound
import org.apache.commons.codec.digest.DigestUtils
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileInputStream
import java.io.IOException

class LocalReadBackend(private val root: File) : StorageReadBackend {

    @Throws(QblStorageException::class)
    override fun download(name: String): StorageDownload {
        try {
            return download(name, null)
        } catch (e: UnmodifiedException) {
            throw IllegalStateException(e)
        }

    }

    @Throws(QblStorageException::class, UnmodifiedException::class)
    override fun download(name: String, ifModifiedVersion: String?): StorageDownload {
        val file = root.resolve(name)

        try {
            if (ifModifiedVersion != null && getMHash(file) == ifModifiedVersion) {
                throw UnmodifiedException()
            }
        } catch (e: IOException) {
            // best effort
        }

        logger.info("Downloading file path " + file)
        try {
            return StorageDownload(
                    FileInputStream(file),
                    getMHash(file),
                    file.length())
        } catch (e: IOException) {
            throw QblStorageNotFound(e)
        }

    }

    @Throws(IOException::class)
    private fun getMHash(file: File): String {
        FileInputStream(file).use { data -> return String(DigestUtils.md5(data)) }
    }

    override fun getUrl(meta: String) = root.resolve(meta).toString()

    companion object {
        private val logger = LoggerFactory.getLogger(LocalReadBackend::class.java)
    }
}
